### ISSUE TITLE
In Safari on iOS 10 beta, map elements disappear when changing filters or rotating map [mitigated]

### ISSUE THREAD

**Issue by carvin** [mapbox/mapbox-gl-js#2986]
**mapbox-gl-js version**: 21.0

This could very well be an issue with the iOS 10 beta and not with Mapbox GL JS, but I wanted to give you folks a heads up.
### Steps to Trigger Behavior
1. Create a page with a Mapbox map
2. Upgrade an iOS device to iOS 10 beta 5 
3. Navigate to the page in Safari and trigger an action that changes the map's filters using _setFilter_. Or, rotate the map using the gesture.
### Behavior

Map elements, with the exception of the labels, disappear or flash. Console warnings that occur when this happens:

```
[Warning] WebGL: INVALID_OPERATION: drawArrays: attempt to access out of bounds arrays (mapbox-gl.js, line 52)
[Warning] WebGL: INVALID_OPERATION: drawArrays: attempt to access out of bounds arrays (mapbox-gl.js, line 32)
```

Here's an example GIF of changing filters on the map:
[(Example Page Here)](http://chrisarv.in/mapbox-bug/)

![filter-bug](https://cloud.githubusercontent.com/assets/550991/17580601/acf3b27a-5f54-11e6-9509-29f48e20957a.gif)

**Comment by rtissera** [comment]
I am also experiencing issues with Safari 10 beta (OS X Sierra DP 5) with correct WebGL code (validated across Firefox and Chrome without any error/warning). This is also failing on a drawArrays call, shall we try to escalate this to Apple ?

**Comment by kristfal** [comment]
🔔  in on this. This issue is still present in `iOS 10 beta 6`. The issue (flickering line and fill styles) occur on all rotate interactions.

The issue is stemming from `drawArrays` in `_renderTileClippingMask`:

```
[Warning] WebGL: INVALID_OPERATION: drawArrays: attempt to access out of bounds arrays (mapbox-gl.js, line XXX)

Offending source line:
```
gl.drawArrays(gl.TRIANGLE_STRIP, 0, this.tileExtentBuffer.length);

```


Just a heads up: iOS is likely to be released to the public on September 7th, which means there is about a week before we'll start seeing a drastic influx of users on iOS10. Would be great if this could be prioritized.

**Comment by lucaswoj** [comment]
I was able to reproduce this bug on my device (iPhone 6, iOS 10 beta 8).

The behavior goes away if the `OES_vertex_array_object` extension is not used ([demo](http://jsbin.com/faniquyudo/edit?html,output))

```js
map.painter.gl.extVertexArrayObject = null;
```

I cannot find any reference to known issues related to vertex array objects in iOS 10. 

The WebGL Vertex Array Object conformance tests on an iPhone 6 running iOS 10 beta 8 and another running iOS 9 produce the same failures (so these failures are unlikely to be related to the root cause of this bug).

```
Testing using buffers that are deleted when attached to bound VAOs

PASS getError was expected value: INVALID_OPERATION : Draw call should fail with unbound position and color buffers
FAIL Position buffer should no longer exist after last ref removed
FAIL Color buffer should no longer exist after last ref removed
PASS getError was expected value: INVALID_OPERATION : Draw call should fail with unbound position and color buffers
FAIL Position buffer should no longer exist after last ref removed
FAIL Color buffer should no longer exist after last ref removed
PASS getError was expected value: INVALID_OPERATION : Draw call should fail with unbound position and color buffers
FAIL Position buffer should no longer exist after last ref removed
FAIL Color buffer should no longer exist after last ref removed
PASS getError was expected value: INVALID_OPERATION : Draw call should fail with unbound position and color buffers
```

# Next Steps

 - [ ] continue looking for issues affecting the vertex array object implementation in iOS 10
 - [ ] consider implementing some browser detection to disable the use of VAOs on iPhones in iOS 10

cc @jfirebaugh @mourner @1ec5 @boundsj @lbud

**Comment by jfirebaugh** [comment]
Is it possible the error message is a true positive and there is actually a bug in our use of VAOs?

**Comment by lucaswoj** [comment]
@jfirebaugh Absolutely. I'll do a close reading of the spec and our code and look for discrepancies.

**Comment by lucaswoj** [comment]
The bug appears iff the map includes a symbol layer. 

```js
    {
        "version": 8,
        "sprite": "mapbox://sprites/mapbox/outdoors-v9",
        "sources": {
            "mapbox://mapbox.mapbox-streets-v7": {
                "url": "mapbox://mapbox.mapbox-streets-v7",
                "type": "vector"
            }
        },
        "layers": [{
            "id": "symbol",
            "type": "symbol",
            "source": "mapbox://mapbox.mapbox-streets-v7",
            "source-layer": "road_label",
            "layout": {
                "icon-image": "marker-15"
            }
        }]
    }
```

**Comment by yeldarby** [comment]
Ouch. This is still an issue on the release version of iOS 10. Its even worse during animations; it causes horrible flickering

**Comment by jordanbommelje** [comment]
Can confirm, still having the issue on iOS 10.0.1 beta, updating a test iPad to iOS 10 public to test as well. Also worth mentioning that this issue is present with Chrome as well as Firefox; tested on a 12" iPad Pro.

**Comment by 1ec5** [comment]
> Also worth mentioning that this issue is present with Chrome as well as Firefox

All browsers on iOS use WebKit, so the issue would manifest itself equally in these browsers too.

**Comment by lucaswoj** [comment]
We added a workaround for this bug in #3207 but have not fully understood the root cause. Keeping this ticket open until we understand the root cause.

**Comment by speedskater** [comment]
@lucaswoj do you have already found the root cause? We had the same issue and in our case the root cause was a call to gl.bindBuffer(gl.ARRAY_BUFFER, null) before gl.drawArrays . Could this be the same issue in your case?

**Comment by lucaswoj** [comment]
Thank you for the tip @speedskater 😄 . This doesn't seem the be the root cause in our case.

**Comment by karhu** [comment]
I've run into a very similar issue in a different, non-mapbox WebGL codebase and while investigating the issue ended up here :) 

Following @speedskater's input, I explicitly rebound the BufferObject containing the vertex data before the affected draw call with `gl.bindBuffer(gl.ARRAY_BUFFER, theBufferObject)` and this seems to solve the issue for us.

Note: We're also working with VAO's, and to my understanding of the WebGL spec, it should not be necessary to have anything bound to to gl.ARRAY_BUFFER_BINDING at the time of the draw call, i.e. the `gl.bindBuffer(gl.ARRAY_BUFFER, null)` call @speedskater mentioned should not pose a problem. This is based on the following statement

> When vertex array pointer state is changed by a call to glVertexAttribPointer, the current buffer object binding (GL_ARRAY_BUFFER_BINDING) is copied into the corresponding client state for the vertex attrib array being changed, one of the indexed GL_VERTEX_ATTRIB_ARRAY_BUFFER_BINDINGs.

from here: https://www.khronos.org/opengles/sdk/docs/man/xhtml/glBindBuffer.xml . 

Long story short, might this be a Safari bug?

**Comment by lucaswoj** [comment]
Thanks @karhu! Like you suggest, we ended up shipping a fix that rebinds the buffer to the VAO more frequently than necessary per the spec. I suspect this is a Safari bug.

**Comment by jkieboom** [comment]
I also had a look at this issue again (in our engine) and I traced it to a bug in WebKit (not only Safari) on drawArrays validation. The bug is that it fails validation on something like the following condition:

```
if (!m_boundArrayBuffer || !m_vertexArrayObject.boundElementArrayBuffer) {
  return false;
}
```

I.e. there must be either an array buffer bound to the gl context, or the VAO _must_ have an element array (index) buffer. Binding either a dummy array buffer to the gl context, or a dummy (1 element sized) element array buffer to the VAO seems to workaround the issue.

Note that it looks like the issue has already been fixed, I can't seem to reproduce the problem in WebKit nightly.

**Comment by mourner** [comment]
@jkieboom thanks for investigating this! Closing the issue because the root cause is found and it's on the WebKit side, and we worked around it in GL JS.
